src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs

```diff
@@ -1004,14 +1007,18 @@ private sealed class AwaitCatchFrame
             private readonly Dictionary<LocalSymbol, LocalSymbol> _hoistedLocals;
             private readonly List<LocalSymbol> _orderedHoistedLocals;
 
-            public AwaitCatchFrame(SyntheticBoundNodeFactory F, TryStatementSyntax tryStatementSyntax)
+            public AwaitCatchFrame(SyntheticBoundNodeFactory F, TryStatementSyntax tryStatementSyntax, AwaitCatchFrame parentOpt = null)
             {
                 this.pendingCaughtException = new SynthesizedLocal(F.CurrentFunction, TypeWithAnnotations.Create(F.SpecialType(SpecialType.System_Object)), SynthesizedLocalKind.TryAwaitPendingCaughtException, tryStatementSyntax);
                 this.pendingCatch = new SynthesizedLocal(F.CurrentFunction, TypeWithAnnotations.Create(F.SpecialType(SpecialType.System_Int32)), SynthesizedLocalKind.TryAwaitPendingCatch, tryStatementSyntax);
 
                 this.handlers = new List<BoundBlock>();
-                _hoistedLocals = new Dictionary<LocalSymbol, LocalSymbol>();
-                _orderedHoistedLocals = new List<LocalSymbol>();
+                _hoistedLocals = parentOpt != null
+                    ? new Dictionary<LocalSymbol, LocalSymbol>(parentOpt._hoistedLocals)
+                    : new Dictionary<LocalSymbol, LocalSymbol>();
+                _orderedHoistedLocals = parentOpt != null
+                    ? new List<LocalSymbol>(parentOpt._orderedHoistedLocals)
```

Comment:
>new List<LocalSymbol>(parentOpt._orderedHoistedLocals) [](http://example.com/codeflow?start=22&length=54)

Is this the right thing to do? It looks like these locals are getting added to a ```BoundBlock``` corresponding to the ```AwaitCatchFrame```. Are we going to add parent locals for all children now? If so, is this intentional?

Reply:
Good catch, thanks. That was indeed wrong. After updating the tests to print output, it even caused a crash in `StackOptimizerPass1`.